### PR TITLE
Node Trace Event prototype

### PR DESCRIPTION
**Pull request by misterpoe**

### CODE REVIEW

File: node.gyp [hunk 1]
```diff
@@ -133,6 +133,7 @@
         'src',
         'tools/msvs/genfiles',
         'deps/uv/src/ares',
+        'deps/v8',
```
Comment (matthewloring) [inline node.gyp]: Can we reduce this dependency or add a TODO. Is it needed for anything other than the trace_event macros?

File: src/node.h [hunk 2]
```diff
@@ -41,6 +41,7 @@
 
 #include "v8.h"  // NOLINT(build/include_order)
 #include "node_version.h"  // NODE_MODULE_VERSION
+#include "src/tracing/trace-event.h"
```
Comment (ofrobots) [inline src/node.h]: Why is this header being included here? It doesn't seem to be getting used in the rest of the file. Would it be better to include the header in the implementation files that actually need it?

File: src/node_tracing_controller.cc [hunk 3]
```diff
@@ -0,0 +1,189 @@
+#include "node_tracing_controller.h"
```
Comment (matthewloring) [inline src/node_tracing_controller.cc]: Can we separate the buffer and writer into their own files (maybe in a tracing directory as you suggested).

Comment (misterpoe) [inline src/node_tracing_controller.cc]: Done

File: src/tracing/trace_event.h [hunk 4]
```diff
@@ -0,0 +1,589 @@
...
+// Set the duration field of a COMPLETE trace event.
+// void TRACE_EVENT_API_UPDATE_TRACE_EVENT_DURATION(
+//     const uint8_t* category_group_enabled,
+//     const char* name,
+//     uint64_t id)
+#define TRACE_EVENT_API_UPDATE_TRACE_EVENT_DURATION             \
+  node::tracing::TraceEventHelper::GetCurrentPlatform() \
+      ->UpdateTraceEventDuration
+
+// Defines atomic operations used internally by the tracing system.
```
Comment (matthewloring) [inline src/tracing/trace_event.h]: Were these non-atomic in the V8 implementation as well?

File: src/tracing/node_trace_buffer.cc [hunk 5]
```diff
@@ -92,5 +84,37 @@ void NodeTraceBuffer::ExtractHandle(
   *event_index = indices % TraceBufferChunk::kChunkSize;
 }
 
+NodeTraceBuffer::NodeTraceBuffer(size_t max_chunks,
+    NodeTraceWriter* trace_writer, Agent* agent)
+    : trace_writer_(trace_writer), current_buf_(0) {
+  for (int i = 0; i < 2; ++i) {
```
Comment (matthewloring) [inline src/tracing/node_trace_buffer.cc]: Prefer to initialize two buffers without loop.

File: src/node_file.cc [hunk 6]
```diff
@@ -325,6 +326,7 @@ static void After(uv_fs_t *req) {
   }
 
   req_wrap->MakeCallback(env->oncomplete_string(), argc, argv);
+  TRACE_EVENT_NESTABLE_ASYNC_END0("node", req_wrap->syscall(), &req_wrap->req_);
```
Comment (ofrobots) [inline src/node_file.cc]: A reminder here that this should be omitted in the initial PR to node and added as a follow-on.

File: src/node.cc [hunk 7]
```diff
@@ -2203,7 +2201,7 @@ static void WaitForInspectorDisconnect(Environment* env) {
 
 void Exit(const FunctionCallbackInfo<Value>& args) {
   WaitForInspectorDisconnect(Environment::GetCurrent(args));
-  v8_platform.Dispose();
+  Environment::GetCurrent(args)->tracing_agent()->Stop();
```
Comment (ofrobots) [inline src/node.cc]: Why are we skipping the disposal of the platform?

File: src/tracing/node_trace_buffer.h [hunk 8]
```diff
@@ -5,6 +5,8 @@
 #include "tracing/node_trace_writer.h"
 #include "libplatform/v8-tracing.h"
 
+#include <atomic>
```
Comment (ofrobots) [inline src/tracing/node_trace_buffer.h]: This is going to be contentious. See: https://chromium-cpp.appspot.com/.

Comment (ofrobots) [inline src/tracing/node_trace_buffer.h]: Some bare-bones atomic support can be used from here: https://github.com/libuv/libuv/blob/b12624c13693c4d29ca84b3556eadc9e9c0936a4/src/unix/atomic-ops.h.

File: src/tracing/trace_event.h (hunk 4 above)
Comment (misterpoe) [inline src/tracing/trace_event.h]: Yes, based on https://cs.chromium.org/chromium/src/v8/src/base/atomicops.h?sq=package:chromium&rcl=1470396217&l=57. TRACE_EVENT_API_ATOMIC_WORD is v8::base::AtomicWord, which is an intptr_t. NoBarrier_Load and NoBarrier_Store are both just standard pointer assignments.

File: src/node.cc (hunk 7 above)
Comment (misterpoe) [inline src/node.cc]: The original Exit did not have the disposal (https://github.com/v8/node/blob/vee-eight-lkgr/src/node.cc). I added it in previously and am now taking it away.

File: src/node_file.cc (hunk 6 above)
Comment (misterpoe) [inline src/node_file.cc]: Added a TODO.

File: src/tracing/node_trace_buffer.cc (hunk 5 above)
Comment (misterpoe) [inline src/tracing/node_trace_buffer.cc]: Done

File: src/node.h (hunk 2 above)
Comment (misterpoe) [inline src/node.h]: Changed to Node's tracing/trace_event.h for global TRACE_EVENT_* macros.

File: node.gyp (hunk 1 above)
Comment (misterpoe) [inline node.gyp]: Done

File: src/tracing/node_trace_buffer.h (hunk 8 above)
Comment (misterpoe) [inline src/tracing/node_trace_buffer.h]: Added a TODO.

**Comment by matthewloring** [comment]
Landed in: https://github.com/matthewloring/node/commit/219c9c64e137da0a466d282b99edafca90998538

**Pull request closed without merging**